### PR TITLE
DDS Security dynamic plugin loading

### PR DESCRIPTION
**Pull request by kurtuluso**
Security plugins are now loaded with the given configuration.
The configuration consists of library-specific properties and security-specific properties.

**Library Specific Properties:**
Path: The file path of the security plugin. It can be either a single library file name without prefix (lib) and suffix (dll/so) or the full absolute file path. Single file is searched in LD_LIBRARY_PATH for POSIX platforms and PATH in the Windows platform.
init function: This is the initialization function that is called just after loading the library. It is expected to initialize the library resources and return the pointer of the plugin object (authentication/access control/cryptography). That object is used in all function calls to the plugin.
finalize function: This is the finalization function that is called before unloading the plugin library. It is expected to release all resources allocated by the plugin library.

**Security specific properties:**
Authentication certificates and private key.
Access control governance and permission file.
These properties will be used after the plugins are ready.

### Requirements
The configuration can be given by either the Cyclone DDS configuration file or ParticipantQoS during participant creation.
ParticipantQoS is required by DDS Security spec. However, the configuration file is an additional useful feature that helps the user to add security to the DDS application without changing the binary.
If ParticipantQoS has a Property starting with the name "dds.sec", then the configuration file is ignored.

The security plugins are loaded with the first participant of the domain and released with the deletion of the domain.

If the security plugin configuration is not given with the Participant QoS, then the XML configuration file is used.
The plugins are loaded only one for the domain by using the configuration of the first participant.

If the participant is the first participant, the security component is initialized. If the participant is the last active participant, then the security component is de-initialized.

**Error handling:**
Plugin configuration and security configuration are checked for lack of property and empty value.
Non-existing plugin library, lack of init and finalize functions, return value of the init function are checked.
If any error occurs, create_participant function returns error code.

### Design:
ddsi_security_omg.c is responsible for the communication and orchestration of DDS Security plugins.
Configuration checking and plugin loading functions are handled in the DDS Security Core component. ddsi_security_omg gets the configuration and ParticipantQoS. After deciding which property set is applicable for security, ddsi_security_omg converts plugin library property set to a complete data structure and Security Core uses this structure for checking and loading purposes.
Loading errors are logged with DDS_ERROR and they cause ddsi_security_omg to return error code.

### Tests
DDSC config tests are reprocessed because the expected participant QoS properties are changed and invalid configurations are not allowed for plugin loading.
These test cases are added:

1. No plugin at the given path
2. No finalize function for the given plugin
3. Failing init function
4. Missing function in auth
5. Missing function in crypto
6. Missing function in ac
7. Successful loading - With configuration.
8. Successful loading - With participant propertyQoS.
9. Missing plugin property in the participant propertyQoS
10. An empty plugin property in the participant propertyQoS
11. Missing security property in the participant propertyQoS
12. Multiple domain configuration

### CODE REVIEW

**Comment by eboasson** [comment]
@kurtuluso, a pull requests on security that changes/adds something while also pulling in tons of commit because master was merged into as well are rather hard to deal with. It is much less problematic to do the merge as a separate pull request (like @martinbremmer just did & that I merged already).

Would you please be so kind as to update this PR to sit on top of current security and do a force push of it so that it contains just the dynamic plugin loading? Thanks!

**Comment by kurtuluso** [comment]
@eboasson now you can review the PR

File: src/security/core/tests/plugin_loading/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+#
+# SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+#
+include (GenerateExportHeader)
+include (CUnit)
+
+set(security_plugin_loading_test_sources
+    "plugin_loading.c"
+  )
+add_definitions(-DDDSI_INCLUDE_SECURITY)
```
[the remark this answers is not in the document]
Comment (kurtuluso) [reply src/security/core/tests/plugin_loading/CMakeLists.txt]: DDSI_INCLUDE_SECURITY is a definition used by the C code. I think you mean ENABLE_SECURITY CMAKE definition which conditionally causes DDSI_INCLUDE_SECURITY to be set. Somehow I needed it for this separate ctest executable but now I moved the DDSI_INCLUDE_SECURITY assignment to security main CMakeList.txt

[the remark this answers is not in the document]
Comment (k0ekk0ek) [reply src/security/core/tests/plugin_loading/CMakeLists.txt]: You're right, I mixed the two up. Still, if that's the case then `DDSI_INCLUDE_SECURITY` should probably be set by the `security_api` target. Then, when it gets linked to the plugin, it'll automatically be propagated. I noticed btw that `add_definitions` is already called in `src/security/CMakeLists.txt` (which is also incorrect). As a general rule, *old-fashioned* CMake usage should be avoided and the target-based approach should be used instead.

**Comment by kurtuluso** [comment]
@eboasson  It is not about the sinking of the messages. I created this PR on Nov 20 and #347 was created on Dec 6. However, you decided to review #347 first.
During the ticket creation, multiple participant configurations are planned to be handled in a separate ticket and at that time, there were no multiple domains. Anyway, now it supports multiple domains.

Code formatting problems mostly occur during merge conflicts. Sorry about empty lines, sometimes I forget them.

About nn_xqos_has_prop, I added check_non_empty flag to this existing function because it was not checking its value but only the name.  startswith and check_non_empty are both true means the name can start with the given name and the value must not be empty. I used your suggestion.

Some of your comments are not acceptable for me to discuss in public. Anyway, presumably, this would be my last PR for the short term.

**Comment by kurtuluso** [comment]
@eboasson One of the builds was broken because of maven repo connectivity problem: https://travis-ci.org/eclipse-cyclonedds/cyclonedds/jobs/634657728?utm_medium=notification&utm_source=github_status

**Comment by eboasson** [comment]
@kurtuluso thank you for updating the pull request.

> @eboasson It is not about the sinking of the messages. I created this PR on Nov 20 and #347 was created on Dec 6. However, you decided to review #347 first.
> During the ticket creation, multiple participant configurations are planned to be handled in a separate ticket and at that time, there were no multiple domains. Anyway, now it supports multiple domains.

I am afraid I don't understand what you are trying to say. Support for multiple domains was introduced by #223 (last August), hence it seems no more than reasonable to expect the loading of security plugins to support that too. Anyway, thanks for sorting that out.

As to ordering, I try to review things mostly in order and certainly I do my best to keep in mind dependencies between PRs when reviewing or merging. Smaller PRs certainly have a better chance of getting reviewed quickly, though. And I do admit to and apologise for my failing to always review promptly.

> Code formatting problems mostly occur during merge conflicts. Sorry about empty lines, sometimes I forget them.

It happens, I know that, and I have no objection to the occasional empty line or a small variation in code formatting here or there. The reason I pointed it out is because of the rather large number of such issues and the recurring nature of them. Even now, after the updates you pushed, they remain ...

> About nn_xqos_has_prop, I added check_non_empty flag to this existing function because it was not checking its value but only the name. startswith and check_non_empty are both true means the name can start with the given name and the value must not be empty. I used your suggestion.
> 
> Some of your comments are not acceptable for me to discuss in public. Anyway, presumably, this would be my last PR for the short term.

Ok, so be it. The saga of this PR combined with this suggests to me that I'd better stop my providing review comments here and try to find another way to get this included. It is an important part of supporting DDS Security, so I guess I'll be reviewing it once more and update it myself.

Thank you for your contributions.

**Comment by eboasson** [comment]
Closed via #372 .

**Pull request closed without merging**